### ISSUE TITLE
State in reducer is probably empty or corrupted even it's not true

### ISSUE THREAD

**Issue by jvaclavik** [reduxjs/redux-toolkit#761]
During the refactoring reducers to createAsyncThunks and normalization from time to time I started to facing the following issue:

`TypeError: Cannot perform 'getPrototypeOf' on a proxy that has been revoked`

![image](https://user-images.githubusercontent.com/858321/95790376-8abfd480-0cdf-11eb-9511-ae3e47c9a3ca.png)

When I refresh the whole page and try to fire this action again, it works correctly, but it looks like that the state is not corrupted while it's not working and has its proper structure. 

I don't understand exactly why it's happening, but it looks like that following reducer `setSelectedTransactionsParams` doesn't get state and it fails: 

**Redux:**

```js
const redux = createSlice({
  name: NAME_PREFIX,
  initialState,
  reducers: {
    setSelectedTransactionsParams: (state, action: PayloadAction<SelectedTransactionsParams>) => ({
      ...state,
      ...{
        selectedTransactionsParams: action.payload,
      }
    })
  },
  extraReducers: {
    // ...
  }
```

**Call from the component:**

```js
  useEffect(() => {
    if (parsedTimeSlotInterval && parsedCategoryIds) {
      dispatch(setSelectedTransactionsParams({
        timeSlotInterval: parsedTimeSlotInterval, categoryIds: parsedCategoryIds, granularity,
      }))
      dispatch(fetchSelectedTransactions())
    }
  }, [])
```
 

Do you have any idea what could cause this problem? Maybe it could cause to immer which is used in redux toolkit.

**Comment by markerikson** [comment]
Huh. Call stack definitely says it's coming from within Immer, but I'd really need to see a reproduction that shows the issue happening to be able to poke at it and get a sense of what's going on.

**Comment by msutkowski** [comment]
I actually ran into this same error the other day when doing a large conversion of an older codebase to use purely RTK. The issue stemmed from having an older immer version in the package-lock _(specifically, I was converting immer 2-based reducers -> createSlice)_. I'd recommend making sure you're not bundling immer yourself, removing your lock file, and doing a reinstall of your node modules. If you're mixing a different immer version with RTK's features, I'd gut any `produce` reference from immer directly, and do:

`import { createNextState as produce } from '@reduxjs/toolkit';`

Note: this is assuming you still have partially migrated things from your refactor. If you don't, I'm probably way off base here :)

**Comment by jvaclavik** [comment]
I probably found how to solve it in my case even though I still don't understand what causes the problem.

I use following dummy function (not very useful in general but it's legacy from older codebase):

```js
export const setValue = (state: any, args: {[key: string]: any}) => (
  { ...state, ...args }
)
```

I didn't use this function in the example above, because I wanted to keep the code simple. 

But when I have it in this state, it's not working:

```js
const redux = createSlice({
  name: NAME_PREFIX,
  initialState,
  reducers: {
    setSelectedTransactionsParams: (state, action: PayloadAction<SelectedTransactionsParams>) => setValue(state, { selectedTransactionsParams: action.payload }),
  },
// ...
```

And this works:

```js
const redux = createSlice({
  name: NAME_PREFIX,
  initialState,
  reducers: {
    setSelectedTransactionsParams: (state, action: PayloadAction<SelectedTransactionsParams>) => ({
      ...state,
      ...{
        selectedTransactionsParams: action.payload,
      }
    })
  },
```

It looks the same to me, but apparently, it makes some difference for immer. 
Am I missing something?

**Comment by jvaclavik** [comment]
I finally found out what was the problem. 
I got an empty array from API and I passed it to `upsertMany` which actually deletes all the entities from the store instead of doing nothing. Is it the right behavior?

See following code:
```js
const redux = createSlice({
  name: NAME_PREFIX,
  initialState,
  reducers: {
    // ...
  },
  extraReducers: (builder) => {
    // ...
    builder.addCase(fetchSelectedTransactions.fulfilled, (state: CurrentState, action: PayloadAction<FetchedTransactions>) => {
      const selectedTransactionIds = action.payload.map((transaction) => transaction.id)
      const newState = setStatus({
        ...state,
        selectedTransactionIds,
      }, Status.SUCCEEDED, fetchSelectedTransactions)
      return transactionsAdapter.upsertMany(newState, action.payload) // the problem is here
    })
    // ...
```

Return should be written like this:

```js
return selectedTransactionIds.length > 0 ? transactionsAdapter.upsertMany(newState, action.payload) : newState
```

btw I use `entities` to store all loaded transactions from API and I pick from the entities from multiple lists of ids (`selectedTransactionIds` is one of them)

**Comment by markerikson** [comment]
That doesn't sound right. Can you create a CodeSandbox that reproduces this issue?

**Comment by msutkowski** [comment]
I was curious about this and made a quick repro and I can't get the described behavior to happen _(it does what you'd expect and not modify the state when passing an empty array)_: https://codesandbox.io/s/test-upsertmany-repro-foeyi?file=/src/features/counter/Counter.tsx. I think it's okay to close this issue.

@markerikson If you'd like, I can throw in a specific test case for this into RTK.

**Comment by markerikson** [comment]
@msutkowski sure, and thanks for looking at it.

**Comment by jvaclavik** [comment]
Yep, I also couldn't replicate this behavior in codesandbox. I will try to explore it later and I will let you know if I make some progress.

**Comment by phryneas** [comment]
@zelenikty the only thing I can see on first look here is that you still have a `    users: usersReducer,` entry in your `configureStore` call, but there is no users slice any more. What exactly are the steps to take to trigger that error?

**Comment by zelenikty** [comment]
I deleted `users: usersReducer` now from the `configureStore `.

Edit:
I just fixed the problem, I just added `selectEntities: selectNotificationsMetadata` selector in the **NotificationsSlice** and now can access all NotificationsMetadata as described in the tutorial.
Before it would break upon having the data fetched from the server successfully and then switching to the Notifications tab in the Navbar.
Adding the `selectEntities: selectNotificationsMetadata` selector is not printed in the code snippets in the tutorial but now it works.
